admin/app/main/s3_client.py

```diff
@@ -1,98 +1,149 @@
 import uuid
 
 import botocore
-from boto3 import resource
+from boto3 import resource, Session
 from flask import current_app
 from notifications_utils.s3 import s3upload as utils_s3upload
 
 FILE_LOCATION_STRUCTURE = 'service-{}-notify/{}.csv'
 TEMP_TAG = 'temp-{user_id}_'
 LOGO_LOCATION_STRUCTURE = '{temp}{unique_id}-{filename}'
 
-def get_s3_object(bucket_name, filename):
-    s3 = resource('s3')
-    return s3.Object(bucket_name, filename)
+class S3Client(object):
+    def __init__(self, bucket_name, session=None):
+        self._bucket_name = bucket_name
+        self._session = session
 
+    @property
+    def session(self):
+        return self._session if self._session else Session()
 
-def delete_s3_object(filename):
-    bucket_name = current_app.config['LOGO_UPLOAD_BUCKET_NAME']
-    get_s3_object(bucket_name, filename).delete()
+    @property
+    def bucket_name(self):
+        return self._bucket_name
 
+    def resource(self, *args, **kwargs):
+        return self.session.resource(*args, **kwargs)
 
-def rename_s3_object(old_name, new_name):
-    bucket_name = current_app.config['LOGO_UPLOAD_BUCKET_NAME']
-    get_s3_object(bucket_name, new_name).copy_from(
-        CopySource='{}/{}'.format(bucket_name, old_name))
-    delete_s3_object(old_name)
+    @property
+    def s3(self):
+        return self.resource('s3')
 
+    @property
+    def bucket(self):
+        return self.s3.Bucket(self._bucket_name)
 
-def get_s3_objects_filter_by_prefix(prefix):
-    bucket_name = current_app.config['LOGO_UPLOAD_BUCKET_NAME']
-    s3 = resource('s3')
-    return s3.Bucket(bucket_name).objects.filter(Prefix=prefix)
+    def get_object(self, filename):
+        return self.s3.Object(self._bucket_name, filename)
 
+    def delete_object(self, filename):
+        return self.get_object(filename).delete()
+
+    def copy_object(self, old_name, new_name):
+        self.get_object(new_name).copy_from(
+            CopySource='{}/{}'.format(self._bucket_name, old_name))
+
+
+class Clients(object):
+    @property
+    def logo(self):
+        return S3Client(
```

Comment:
this might need some micro-optimization.

s3 boto library re-uses s3 sessions instead of re-instantiating them

Reply:
see here

https://github.com/boto/boto3/blob/develop/boto3/__init__.py#L72